core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -0,0 +1,378 @@
+        assert accumulateJoins.get() : "waitToBeElectedAsMaster is called we are not accumulating joins";
+
+        final CountDownLatch done = new CountDownLatch(1);
+        final ElectionContext newContext = new ElectionContext(callback, requiredMasterJoins) {
+            @Override
+            void onClose() {
+                if (electionContext.compareAndSet(this, null)) {
+                    stopAccumulatingJoins();
+                } else {
+                    assert false : "failed to remove current election context";
+                }
+                done.countDown();
+            }
+        };
+
+        if (electionContext.compareAndSet(null, newContext) == false) {
+            // should never happen, but be conservative
+            callback.onFailure(new IllegalStateException("double waiting for election"));
+            return;
+        }
+        try {
+            // check what we have so far..
+            checkPendingJoinsAndElectIfNeeded();
+
+            try {
+                if (done.await(timeValue.millis(), TimeUnit.MILLISECONDS)) {
+                    // callback handles everything
+                    return;
+                }
+            } catch (InterruptedException e) {
+
+            }
+            if (logger.isTraceEnabled()) {
+                final int pendingNodes;
+                synchronized (pendingJoinRequests) {
+                    pendingNodes = pendingJoinRequests.size();
+                }
+                logger.trace("timed out waiting to be elected. waited [{}]. pending node joins [{}]", timeValue, pendingNodes);
+            }
+            // callback will clear the context, if it's active
+            newContext.onFailure(new ElasticsearchTimeoutException("timed out waiting to be elected"));
+        } catch (Throwable t) {
+            logger.error("unexpected failure while waiting for incoming joins", t);
+            newContext.onFailure(t);
+        }
+    }
+
+    public void startAccumulatingJoins() {
+        logger.trace("starting to accumulate joins");
+        boolean b = accumulateJoins.getAndSet(true);
+        assert b == false : "double startAccumulatingJoins() calls";
+        assert electionContext.get() == null : "startAccumulatingJoins() called, but there is an ongoing election context";
+    }
+
+    public void stopAccumulatingJoins() {
+        logger.trace("stopping join accumulation");
+        assert electionContext.get() == null : "stopAccumulatingJoins() called, but there is an ongoing election context";
+        boolean b = accumulateJoins.getAndSet(false);
+        assert b : "stopAccumulatingJoins() called but not accumulating";
+        synchronized (pendingJoinRequests) {
+            if (pendingJoinRequests.size() > 0) {
+                processJoins("stopping to accumulate joins");
+            }
+        }
+    }
+
+    public void handleJoinRequest(final DiscoveryNode node, final MembershipAction.JoinCallback callback) {
```

Comment:
true, the node trying join, will then just try again.

What I was wondering about is, if there are any downsides with waiting for return the join request calls? Obviously there is a big upside, that is why this PR exists :) but I'm trying to think about any downsides I'm maybe overlooking, because prior to this change we used return immediately.